test/multi_rollup_sequencer.go

```diff
@@ -0,0 +1,134 @@
+package test
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/rollkit/go-sequencing"
+)
+
+// MultiRollupSequencer is a sequencer for testing that serves multiple rollups
+type MultiRollupSequencer struct {
+	rollups      map[string]*RollupData
+	rollupsMutex sync.RWMutex
+}
+
+// RollupData holds the data for a specific rollup, including its transaction queue, last batch hash, and seen batches.
+type RollupData struct {
+	tq                 *TransactionQueue
+	lastBatchHash      []byte
+	lastBatchHashMutex sync.RWMutex
+
+	seenBatches      map[string]struct{}
+	seenBatchesMutex sync.Mutex
+}
+
+// SubmitRollupTransaction implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) SubmitRollupTransaction(ctx context.Context, req sequencing.SubmitRollupTransactionRequest) (*sequencing.SubmitRollupTransactionResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+	rollup.tq.AddTransaction(req.Tx)
+	return &sequencing.SubmitRollupTransactionResponse{}, nil
+}
+
+// GetNextBatch implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) GetNextBatch(ctx context.Context, req sequencing.GetNextBatchRequest) (*sequencing.GetNextBatchResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	rollup.lastBatchHashMutex.RLock()
+	lastBatchHash := rollup.lastBatchHash
+	rollup.lastBatchHashMutex.RUnlock()
+
+	if lastBatchHash == nil && req.LastBatchHash != nil {
+		return nil, errors.New("lastBatch is supposed to be nil")
+	} else if lastBatchHash != nil && req.LastBatchHash == nil {
+		return nil, errors.New("lastBatch is not supposed to be nil")
+	} else if !bytes.Equal(lastBatchHash, req.LastBatchHash) {
+		return nil, errors.New("supplied lastBatch does not match with sequencer last batch")
+	}
+
+	batch := rollup.tq.GetNextBatch(req.MaxBytes)
+	batchRes := &sequencing.GetNextBatchResponse{Batch: batch, Timestamp: now}
+	// If there are no transactions, return empty batch without updating the last batch hash
+	if len(batch.Transactions) == 0 {
+		return batchRes, nil
+	}
+
+	h, err := batch.Hash()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Shouldn't an error with hashing the batch should be check prior to returning `batchRes` if there are no transactions? Otherwise you are potentially returning an error prone batch within the batch response.

Reply:
not needed as the receiver only hashes if there are transactions.